### ISSUE TITLE
[BUG] Issue retrieving secret resource in container in Istio environment

### ISSUE THREAD

**Issue by Aaron-ML** [SparebankenVest/azure-key-vault-to-kubernetes#34]
**Note: Make sure to check out known issues (https://github.com/sparebankenvest/azure-key-vault-to-kubernetes#known-issues) before submitting**

**Describe the bug**
I'm able to get the init-container to startup, I can see it getting mutated in the akvs pod logs, however in the **Container log**:

`level=fatal msg="env-injector: error getting azurekeyvaultsecret resource 'password', error: an error on the server (\"\") has prevented the request from succeeding (get azurekeyvaultsecrets.spv.no password)"`

**To Reproduce**

**Kubernetes Version:** v1.15.7

**AKVS Definition:**
```
apiVersion: spv.no/v1
kind: AzureKeyVaultSecret
metadata:
  name: password
  namespace: default
spec:
  vault:
    name: vault
    object:
      name: password
      type: secret
```
**Namespace label:**
```
apiVersion: v1
kind: Namespace
metadata:
  labels:
    azure-key-vault-env-injection: enabled
    istio-injection: enabled
  name: default
spec:
  finalizers:
  - kubernetes
```

**Expected behavior**
I expect the env variable to get injected from the azure key vault
**Logs**

**Env Injector Pod logs**
```
time="2020-02-21T03:23:32Z" level=info msg="checking for env vars containing '@azurekeyvault' in container stats-api"
time="2020-02-21T03:23:32Z" level=info msg="found env var: password@azurekeyvault"
time="2020-02-21T03:23:32Z" level=info msg="found credentials to use with registry 'repo.azurecr.io'"
time="2020-02-21T03:23:32Z" level=info msg="pulling docker image epo.azurecr.io/com.company/app:b327a2to get entrypoint and cmd, timeout is 120 seconds"
time="2020-02-21T03:23:33Z" level=info msg="docker image epo.azurecr.io/com.company/app:b327a2 pulled successfully"
time="2020-02-21T03:23:33Z" level=info msg="inspecting container image repo.azurecr.io/com.company/app:b327a2, looking for entrypoint and cmd"
time="2020-02-21T03:23:33Z" level=info msg="using cmd from image: [sh -c java $JAVA_OPTS -jar /usr/src/service/service.jar]"
time="2020-02-21T03:23:33Z" level=info msg="using cmd from image: [sh -c java $JAVA_OPTS -jar /usr/src/service/service.jar]"
time="2020-02-21T03:23:33Z" level=info msg="using 'sh -c java $JAVA_OPTS -jar /usr/src/service/service.jar' as arguments for env-injector"
time="2020-02-21T03:23:33Z" level=info msg="containers mutated and pod updated with init-container and volumes"
```
**Additional context**
We are running istio which has it's own init container, but I've tried with the annotation to allow all ingress, along with destinationrules and serviceentries to allow anything to the keyvaults. I don't see anything on Istio that would be blocking this.

**Comment by torresdal** [comment]
@Aaron-ML Have you tried to increase the env-injector log-level on your Pod - to get a bit more info about what's going on? https://akv2k8s.io/troubleshooting

**Comment by Aaron-ML** [comment]
@torresdal Yes, I've enabled debug mode but I'm not seeing anything overly suspect.

I did find however that the pods did not have access to the AKVS resource from the API-Server, which was odd considering the RBAC policies exist. 

It appears to be somewhat working now, the next issue I'm facing is that on bootup the pods will crashloop once before they pull the secret successfully, I'm not sure if it's a probe failing or if it's related to pulling the image.. I'm working on getting more output for that.

**Comment by Aaron-ML** [comment]
This issue appeared to be related to istio mTLS and how AKS propagates the external API.

I needed to add a destinationRule specifically for 'kubernetes' and not the external IP or 'kubernetes.default.svc' to disable mTLS requirements. 

```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  annotations:
  name: akvs-external
  namespace: default
spec:
  host: kubernetes
  trafficPolicy:
    tls:
      mode: DISABLE
```

The next issue I'm troubleshooting is the pods restarting 1-2 times before running without issues. @torresdal not sure if that's normal behavior or not. 

akvs pod logs:
```
time="2020-02-29T03:27:29Z" level=info msg="found pod to mutate in namespace 'default'"
time="2020-02-29T03:27:29Z" level=info msg="found container 'app' to mutate"
time="2020-02-29T03:27:29Z" level=info msg="checking for env vars containing '@azurekeyvault' in container app"
time="2020-02-29T03:27:29Z" level=info msg="found env var: password@azurekeyvault"
time="2020-02-29T03:27:29Z" level=info msg="found credentials to use with registry 'repo.azurecr.io'"
time="2020-02-29T03:27:29Z" level=info msg="pulling docker image repo.azurecr.io/com.repo/app:eba0e64 to get entrypoint and cmd, timeout is 120 seconds"
time="2020-02-29T03:27:29Z" level=info msg="docker image repo.azurecr.io/com.repo/app:eba0e64 pulled successfully"
time="2020-02-29T03:27:29Z" level=info msg="inspecting container image repo.azurecr.io/com.repo/app:eba0e64, looking for entrypoint and cmd"
time="2020-02-29T03:27:29Z" level=info msg="using cmd from image: [sh -c java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar]"
time="2020-02-29T03:27:29Z" level=info msg="using cmd from image: [sh -c java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar]"
time="2020-02-29T03:27:29Z" level=info msg="using 'sh -c java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar sh -c java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar' as arguments for env-injector"
time="2020-02-29T03:27:29Z" level=info msg="containers mutated and pod updated with init-container and volumes"
2020/02/29 03:27:29 [DEBUG] json patch for request 23d056cc-a985-4d7c-a620-5432cfbf7fcc: [{"op":"add","path":"/spec/containers/0/args","value":["sh","-c","java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar","sh","-c","java $JAVA_OPTS -Dlog4j.configurationFile=log4j2-common.xml -Dvertx.logger-delegate-factory-class-name=io.vertx.core.logging.Log4j2LogDelegateFactory -jar /usr/src/service/service.jar"]},{"op":"add","path":"/spec/containers/0/volumeMounts/2","value":{"mountPath":"/azure-keyvault/","name":"azure-keyvault-env"}},{"op":"add","path":"/spec/containers/0/command","value":["/azure-keyvault/azure-keyvault-env"]},{"op":"add","path":"/spec/containers/0/env/5","value":{"name":"ENV_INJECTOR_CUSTOM_AUTH","value":"false"}},{"op":"add","path":"/spec/containers/0/env/5","value":{"name":"ENV_INJECTOR_POD_NAMESPACE","valueFrom":{"fieldRef":{"fieldPath":"metadata.namespace"}}}},{"op":"add","path":"/spec/initContainers","value":[{"command":["sh","-c","cp /usr/local/bin/azure-keyvault-env /azure-keyvault/ \u0026\u0026 cp /etc/kubernetes/azure.json /azure-keyvault/azure.json \u0026\u0026 chmod 444 /azure-keyvault/azure.json"],"image":"spvest/azure-keyvault-env:0.1.15","imagePullPolicy":"IfNotPresent","name":"copy-azurekeyvault-env","resources":{},"volumeMounts":[{"mountPath":"/azure-keyvault/","name":"azure-keyvault-env"},{"mountPath":"/etc/kubernetes/azure.json","name":"azure-config","readOnly":true}]}]},{"op":"add","path":"/spec/volumes/2","value":{"emptyDir":{"medium":"Memory"},"name":"azure-keyvault-env"}},{"op":"add","path":"/spec/volumes/3","value":{"hostPath":{"path":"/etc/kubernetes/azure.json","type":"File"},"name":"azure-config"}}]
```

kubectl get pods
```
NAME                           READY  STATUS   RESTARTS   AGE
app-5cfdb5785c-k757w            2/2  Running   2    5m

**Comment by torresdal** [comment]
@Aaron-ML  Thanks for detailing this issue. Was not aware of issues related to Istio, so finding solutions and workarounds is appreciated!

The Env-Injector log looks normal, the restarts are not. What does `kubectl describe` say with regards to restart reason for your `app-5cfdb5785c-k757w ` Pod?

**Comment by Aaron-ML** [comment]
I was able to grab the error in the pod when it fails:
```
level=fatal msg="env-injector: error getting azurekeyvaultsecret resource 'password', error: Get https://10.0.0.1:443/apis/spv.no/v1alpha1/namespaces/default/azurekeyvaultsecrets/password: dial tcp 10.0.0.1:443: connect: connection refused"
```

I believe I'm running into an issue where istio-proxy is setting up and switching to proxy the network on the pod, and the azure-envinjector container crashes during this change as they both run at the same time. Not quite sure how to fix this other than to introduce some delay on the env-injector container.

**Comment by torresdal** [comment]
That it fails and restarts is not ideal I agree, but at the same time it does this for a reason and manage to recover eventually. As you say, I'm not sure either how we would go about fixing this, so I suggest we leave it for now.

What would be great is if you could take the time to summarize this problem and your solution in a pull-request for the Troubleshooting docs: https://akv2k8s.io/troubleshooting - in github: https://github.com/SparebankenVest/azure-key-vault-to-kubernetes/blob/master/docs/content/troubleshooting.md

**Comment by torresdal** [comment]
You should also consider reporting this as a bug to Istio.

**Comment by torresdal** [comment]
@Aaron-ML I added a commit which might solve this for you: https://github.com/SparebankenVest/azure-key-vault-to-kubernetes/commit/8d2b92c823765fcdc1ad28cd8885a80225dcf996

Added default 3 retries, waiting 3 seconds between retries. These values are configurable through these env vars: ENV_INJECTOR_RETRIES and ENV_INJECTOR_WAIT_BEFORE_RETRY.

Let you know as soon as there are container images available you can try out.

**Comment by Aaron-ML** [comment]
@torresdal Thanks!, I think that should help for sure.

I'm hoping this Issue gets resolved eventually.. looks like sidecar logic has been in the kubernetes pipeline for awhile now. https://github.com/kubernetes/enhancements/issues/753

**Comment by Aaron-ML** [comment]
@toreandreb Version 1.0.1 looks better, I do occasionally get a CrashLoop on startup of pods with the following error:
`
Warning  Failed                  13m                 kubelet, aks-agentpool-34239724-vmss00002q  Error: failed to start container "stats-api": Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "exec: \"/azure-keyvault/azure-keyvault-env\": stat /azure-keyvault/azure-keyvault-env: no such file or directory": unknown
`

**Comment by torresdal** [comment]
@Aaron-ML regarding your last error - see #42

**Comment by MarianZoll** [comment]
@Aaron-ML 

We fixed this with Istio 1.7 by applying the istio proxy sidecar dependency as described here: https://github.com/istio/istio/pull/24737 

Additionally, we ensured that both `istio-init` and `istio-proxy` have at least 200m CPU and 200Mi memory as pod limits. Without those, the `istio-pilot` will exit in 137. 

Hopefully this will find someone else looking for a solution.
